### PR TITLE
Fix race condition in NaiveEngine::PushAsync

### PR DESCRIPTION
**Pull request by leezu**
Wait for async_fun to complete in NaiveEngine::PushAsync.

This fixes a race condition in which NaiveEngine::PushAsync was checking if the
the async_fun had completed by the end of NaiveEngine::PushAsync scope.

If async_fun hadn't completed yet, NaiveEngine::PushAsync would set an internal error string
and deallocate the callback, causing segfault in async_fun once it would attempt
calling the callback.

### CODE REVIEW

**Comment by leezu** [comment]
@mxnet-bot run ci [all]

**Comment by leezu** [comment]
@mxnet-bot run ci [all]

File: src/engine/naive_engine.cc [hunk 1]
```diff
@@ -208,8 +210,7 @@ class NaiveEngine final : public Engine {
     for (auto var : mutable_vars) {
       ++var->version_;
     }
-    CHECK(req_completed)
-        << "NaiveEngine only support synchronize Push so far";
+    future.wait();
```
Comment (ptrendx) [inline src/engine/naive_engine.cc:213]: With naive engine it does not really make a difference but generally speaking this wait should be before the `version_` increment, right?

**Review by ptrendx — APPROVED** [review]
LGTM

**Comment by leezu** [comment]
@mxnet-bot run ci [windows-gpu]

**Comment by leezu** [comment]
https://github.com/thrust/thrust/issues/1090 comes back with high frequency in this PR... 10 compilation attempts in a row failed (each CI run attempts compilation 5 times before giving up).

**Comment by leezu** [comment]
@mxnet-bot run ci [windows-gpu]

**Comment by leezu** [comment]
@mxnet-bot run ci [windows-gpu]

**Pull request merged**